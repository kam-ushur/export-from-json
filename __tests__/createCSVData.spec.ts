import { testTableItems, testTableItems2, testTableItems3, beforeTableEncode } from './shared/data'
import { createCSVData } from '../src/converters'

test('createCSVData works correctly', () => {
  expect(createCSVData(testTableItems)).toMatchSnapshot()
  expect(createCSVData(testTableItems2)).toMatchSnapshot()
})

test('createCSVData works correctly with specifying beforeTableEncode option', () => {
  expect(createCSVData(testTableItems, { beforeTableEncode })).toMatchSnapshot()
})
test('createCSVData works correctly with delimiter option', () => {
  expect(createCSVData(testTableItems, { delimiter: ';' })).toMatchSnapshot()
})

test('createCSVData works correctly with specifying sanitizeCell option', () => {
  function sanitizeCell(value: string, delimiter: ',' | ';') {
    const enclosingTester = new RegExp(`${delimiter}|"|\n`)
    // Check if the field requires wrapping
    const requiresWrapping = enclosingTester.test(value) || /^[=+\-@]/.test(value);
  
    if (requiresWrapping) {
      // Escape double quotes by replacing each " with ""
      const escaped = value.replace(/"/g, '""');
      // Finally wrap in double quotes
      return `"${escaped}"`;
    }
  
    return value;
  }
  expect(createCSVData(testTableItems3, { sanitizeCell })).toMatchSnapshot()
})


test('createCSVData works with empty array', () => {
  expect(createCSVData([])).toEqual('')
})
