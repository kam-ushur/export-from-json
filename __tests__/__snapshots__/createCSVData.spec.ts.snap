// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCSVData works correctly 1`] = `
"greeting,foo,bar,a,b,c,d,e,f
hello,,,,,,,,
,foo,bar,,,,,,
,,,100,"[1,2,3]","{""x"":200,""y"":300}",D,null,"
`;

exports[`createCSVData works correctly 2`] = `
"greeting,foo,bar,a,b,c,d,e,f
"hel""lo",,,,,,,,
,foo,"b""a""r",,,,,,
,,,100,"[1,2,3]","{""x"":200,""y"":300}",D,null,"
`;

exports[`createCSVData works correctly with delimiter option 1`] = `
"greeting;foo;bar;a;b;c;d;e;f
hello;;;;;;;;
;foo;bar;;;;;;
;;;100;[1,2,3];"{""x"":200,""y"":300}";D;null;"
`;

exports[`createCSVData works correctly with specifying beforeTableEncode option 1`] = `
"greeting,foo,bar,a,b,c,d,e,f
hello,0,0,0,0,0,0,0,0
0,foo,bar,0,0,0,0,0,0
0,0,0,1000,"[1,2,3]","{""x"":200,""y"":300}",D,null,0"
`;

exports[`createCSVData works correctly with specifying sanitizeCell option 1`] = `
"greeting,morequotes,simple,"he""a,der",foo,bar,fizz,buzz,complex1,complex2,formula1,formula2,jab,duck,uppercut,kick,a,b,c,d,e,f
"hel""lo","""already quoted""",simple text,some header,,,,,,,,,,,,,,,,,,
,,,,"foo, thing","b""a""r",,,,,,,,,,,,,,,,
,,,,,,"some
thing","b""a""r",,,,,,,,,,,,,,
,,,,,,,,"=Hello, ""world""
@2+2","b""a""r","""""""=2+5+cmd|'/C calc'!A0""""""","=1+2"";=1+2",,,,,,,,,,
,,,,,,,,,,,,"=4523","+12","-12","@user",,,,,,
,,,,,,,,,,,,,,,,100,"[1,2,3]","{""x"":200,""y"":300}",D,null,"
`;
